contracts/src/BeefyClient.sol

```diff
@@ -390,49 +417,57 @@ contract BeefyClient {
+    // signature was used.
+    //
+    // ceil(log2(validatorSetLen)) + 1 * 2 ceil(log2(signatureUseCount))
+    //
+    // See https://hackmd.io/9OedC7icR5m-in_moUZ_WQ for full analysis.
+    function signatureSamples(uint256 validatorSetLen, uint256 signatureUseCount) internal view returns (uint256) {
+        // There are less validators than the minimum signatures so validate 2/3 majority.
+        if (validatorSetLen <= minimumSignatureSamples) {
             return validatorSetLen - (validatorSetLen - 1) / 3;
-        } else if (validatorSetLen < 342) {
-            return 10;
-        } else if (validatorSetLen < 683) {
-            return 11;
-        } else if (validatorSetLen < 1366) {
-            return 12;
-        } else if (validatorSetLen < 2731) {
-            return 13;
-        } else if (validatorSetLen < 5462) {
-            return 14;
-        } else if (validatorSetLen < 10923) {
-            return 15;
-        } else if (validatorSetLen < 21846) {
-            return 16;
-        } else {
-            return 17;
         }
+
+        // Start with the minimum number of signatures.
+        uint256 samples = minimumSignatureSamples;
+
+        // We must substrate minimumSignatures from the number of validators or we might end up
+        // requiring more signatures than there are validators.
+        uint256 extraValidators = validatorSetLen - minimumSignatureSamples;
+        // Using highest bit set to yield floor(log2(x))
+        uint256 validatorsLog2 = Bits.highestBitSet(extraValidators);
+        // We require ceil(log2(x)) so we round up if there were any extra bits set.
+        if (extraValidators > (1 << validatorsLog2)) {
+            validatorsLog2++;
+        }
+        samples += validatorsLog2;
+
+        // No signatures use count. Just return samples.
+        if (signatureUseCount == 0) {
+            return samples;
+        }
+
+        // To address the concurrency issue specified in the link below:
+        // https://hackmd.io/wsVcL0tZQA-Ks3b5KJilFQ?view#Solution-2-Signature-Checks-dynamically-depend-on-the-No-of-initial-Claims-per-session
+        // It must be harder for a mallicious relayer to spam submitInitial to bias the RANDAO.
+        // If we detect that a signature is used many times (spam), we increase the number of signature samples required on submitFinal.
+
+        // Using highest bit set to yield floor(log2(x))
+        uint256 log2SignatureUse = Bits.highestBitSet(signatureUseCount);
+        // We require ceil(log2(x)) so we round up if there were any extra bits set.
+        if (signatureUseCount > (1 << log2SignatureUse)) {
+            ++log2SignatureUse;
+        }
+        // Based on formula provided here: https://hackmd.io/9OedC7icR5m-in_moUZ_WQ
+        samples += 1 + 2 * log2SignatureUse;
+
+        // Make sure that samples are bounded by the validatorSetLength
+        if (samples > validatorSetLen) {
+            samples = validatorSetLen;
+        }
```

Comment:
Also known as `Math.min(x, y)`.

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/956/commits/51746e56272d3a94dab96546e0422abbd0b33826